drafts/element-internals-api.md

```diff
@@ -0,0 +1,179 @@
+---
+title: Finally, custom form elements that don't suck!
+domain: campzulu.hashnode.dev
+tags: web-development, lit, web-components, shadow-dom, form
+cover: https://cdn.hashnode.com/res/hashnode/image/stock/unsplash/jJT2r2n7lYA/upload/76adc03462239d3b3e6cb53f44b60b20.jpeg?w=1600&h=840&fit=crop&crop=entropy&auto=compress,format&format=webp
+publishAs: r3dDoX
+hideFromHashnodeCommunity: false
+ignorePost: true
+---
+
+In the world of web development, creating custom form elements has always been a bit of a challenge.
+However, with the introduction of web components and the use of the ElementInternals API, this task
+has become much easier. By leveraging the power of web components, developers can now create custom
+form elements that can be used in any website or application. In this blog post, we will explore
+what the ElementInternals API is and how you can use it to create custom form elements.
+
+## The Problem
+
+One of the main challenges that developers face when creating web components is making them work
+seamlessly with HTML forms. By default, web components are not recognized as form elements, which
+means that they cannot be submitted and don't show up in FormData. So even if your custom component
+wraps a native input in its ShadowDOM, it will not be picked up by a form outside the Shadow DOM.
+You could pass a native input into the component from the outside via the `slot` element, but this
+creates numerous new problems with accessing and styling that input from inside the web component.
+This also means they won't work out of the box with common form libraries.
+
+## ElementInternals to the rescue
+
+This [API](https://developer.mozilla.org/en-US/docs/Web/API/ElementInternals) provides a set of
+methods and properties that allow developers to customize the behavior of web components when used
+as form elements. It is supported by all major browsers, including Chrome ( v77), Firefox (v93),
+Safari (v16.4), and Edge (v79). See [caniuse](https://caniuse.com/mdn-api_elementinternals).
+If you need to support older browsers there is a polyfill with limited functionality available:
+https://www.npmjs.com/package/element-internals-polyfill.
+
+As there are numerous properties and methods available, let's focus on the most significant ones:
+
+* Start using ElementInternals in your web component
+* Ensure value and validity
+* Focus events
+
+## Start using ElementInternals in your web component
+
+There is essentially two things we have to do, to use the ElementInternals API:
+
+* Flag the web component as being `formAssociated`
+* Inject the ElementInterals in the constructor of our web component
```

Comment:
I think this step could be described a bit more in detail. With "Inject" I associated constructor injection, i.e. that it gets passed as a parameter to the constructor. However below in the code sample I learned that we actually call the .attachInternals function to get the internals.

Reply:
I called it "attach" now. This way we use the same naming. There is not much more details to go into without going into Browser internals. So I hope this makes it clear enough